replace siphash128 with foldhash::quality

conradludgate opened the pull request:
As the title suggests, this PR replaces the siphash based phf with a foldhash based phf. [foldhash](https://docs.rs/foldhash/latest/foldhash/) is a new fast hash algorithm by @orlp which has high avalanche properties and is very fast on modern hardware. My benchmarks demonstrate that construction times are similar (sometimes faster, sometimes slower - only a few % difference), but lookup times are significantly improved (2.5x).

To get more reliable benchmarks, I had to modify them. 
* The construction benchmarks were biased towards lucky seeds, so I made them change seed each time.
* The lookup benchmarks for hashmap/btreemap were unstable based on the insertion order, so I fixed the order to make it more consistent.

<details>
<summary>Benchmark reports</summary>

These benchmarks were run on my M2 Max.

## Construction benchmark results:

This is the output with `foldhash::quality`, against a baseline with `siphash`

```
gen_hash_small/0        time:   [38.255 ns 38.407 ns 38.566 ns]
                        change: [-7.3965% -6.6754% -5.9733%] (p = 0.00 < 0.05)

gen_hash_small/1        time:   [137.59 ns 137.88 ns 138.20 ns]
                        change: [+1.7959% +2.2170% +2.6189%] (p = 0.00 < 0.05)

gen_hash_small/2        time:   [177.49 ns 177.91 ns 178.40 ns]
                        change: [-6.4105% -6.0836% -5.7827%] (p = 0.00 < 0.05)

gen_hash_small/5        time:   [1.4264 µs 1.4298 µs 1.4337 µs]
                        change: [-10.587% -10.189% -9.7465%] (p = 0.00 < 0.05)

gen_hash_small/10       time:   [9.5154 µs 9.5789 µs 9.6682 µs]
                        change: [-1.4397% -0.3968% +0.8412%] (p = 0.52 > 0.05)

gen_hash_small/25       time:   [35.824 µs 35.946 µs 36.063 µs]
                        change: [-5.5198% -4.2950% -3.0867%] (p = 0.00 < 0.05)

gen_hash_small/50       time:   [83.022 µs 83.504 µs 84.256 µs]
                        change: [+0.4777% +2.0264% +3.7782%] (p = 0.01 < 0.05)

gen_hash_small/75       time:   [88.571 µs 89.456 µs 90.760 µs]
                        change: [+8.6597% +9.7361% +10.822%] (p = 0.00 < 0.05)

gen_hash_medium/100     time:   [139.03 µs 140.40 µs 142.92 µs]
                        change: [+1.7471% +3.0430% +4.3443%] (p = 0.00 < 0.05)

gen_hash_medium/250     time:   [242.82 µs 245.76 µs 249.84 µs]
                        change: [+4.3653% +6.1695% +8.3223%] (p = 0.00 < 0.05)

gen_hash_medium/500     time:   [423.64 µs 432.40 µs 445.72 µs]
                        change: [-5.8532% -4.2099% -2.3072%] (p = 0.00 < 0.05)

gen_hash_medium/1000    time:   [751.80 µs 754.20 µs 756.92 µs]
                        change: [-9.4598% -8.1575% -6.8389%] (p = 0.00 < 0.05)

gen_hash_medium/2500    time:   [1.7661 ms 1.7771 ms 1.7923 ms]
                        change: [-2.4366% +0.1787% +2.0336%] (p = 0.90 > 0.05)

gen_hash_medium/5000    time:   [3.6819 ms 3.6914 ms 3.7013 ms]
                        change: [+3.2364% +4.7253% +5.7647%] (p = 0.00 < 0.05)
                        
gen_hash_medium/7500    time:   [5.4155 ms 5.4507 ms 5.5126 ms]
                        change: [-5.4982% -4.2662% -2.8330%] (p = 0.00 < 0.05)

gen_hash_large/10000    time:   [7.3447 ms 7.3633 ms 7.3841 ms]
                        change: [-4.5368% -2.6607% -1.0655%] (p = 0.00 < 0.05)

gen_hash_large/25000    time:   [19.404 ms 19.448 ms 19.500 ms]
                        change: [-2.3404% -0.8883% +0.2431%] (p = 0.21 > 0.05)

gen_hash_large/50000    time:   [41.211 ms 41.273 ms 41.344 ms]
                        change: [+2.6534% +2.9329% +3.2080%] (p = 0.00 < 0.05)

gen_hash_large/75000    time:   [62.886 ms 63.090 ms 63.322 ms]
                        change: [-5.1691% -4.7559% -4.3487%] (p = 0.00 < 0.05)

gen_hash_xlarge/100000  time:   [86.515 ms 86.861 ms 87.244 ms]
                        change: [+3.6388% +4.0567% +4.5285%] (p = 0.00 < 0.05)

gen_hash_xlarge/250000  time:   [287.58 ms 288.54 ms 289.63 ms]
                        change: [+22.519% +23.024% +23.549%] (p = 0.00 < 0.05)

gen_hash_xlarge/500000  time:   [523.20 ms 525.07 ms 527.37 ms]
                        change: [-5.8478% -5.4520% -5.0033%] (p = 0.00 < 0.05)

gen_hash_xlarge/750000  time:   [1.2349 s 1.2373 s 1.2399 s]
                        change: [-1.6023% -1.0373% -0.4986%] (p = 0.00 < 0.05)

gen_hash_xlarge/1000000 time:   [1.9449 s 1.9529 s 1.9615 s]
                        change: [+19.684% +20.257% +20.829%] (p = 0.00 < 0.05)
```

## Lookup benchmark results:

Baseline:

```
test map::bench_btreemap_none ... bench:          14.17 ns/iter (+/- 0.41)
test map::bench_btreemap_some ... bench:          19.98 ns/iter (+/- 1.08)
test map::bench_hashmap_none  ... bench:           8.06 ns/iter (+/- 0.27)
test map::bench_hashmap_some  ... bench:          17.77 ns/iter (+/- 2.95)
test map::bench_match_none    ... bench:           2.71 ns/iter (+/- 0.19)
test map::bench_match_some    ... bench:           2.88 ns/iter (+/- 0.07)
```

Siphash:

```
test map::bench_phf_none      ... bench:          10.28 ns/iter (+/- 0.11)
test map::bench_phf_some      ... bench:          12.52 ns/iter (+/- 0.21)
```

foldhash:

```
test map::bench_phf_none      ... bench:           4.03 ns/iter (+/- 0.04)
test map::bench_phf_some      ... bench:           5.13 ns/iter (+/- 0.04)
```

</details>

conradludgate commented:
Well, one of the tests is failing. That's odd... and concerning

conradludgate commented:
Annoying case with `unicase`. 
* `UniCase::ascii("foo")` is hashed as `b'f', b'o', b'o'`
* `UniCase::unicode("foo")` is hashed as `b"f", b"o", b"o"`

In SipHash, these end up being equivalent (since siphash is a bytestream digest), but it's different in foldhash.

JohnTitor commented:
Thanks! We've switched to fastrand in #312, could you check if it's still effective?

conradludgate commented:
> could you check if it's still effective?

Construction time seems to be the same (+-5% in the majority of tests).
Lookup time is still 2.5x faster

conradludgate commented:
Tested some more using foldhash::fast, and it seems to still have good performance (slightly better), so I switched to that and removed the code that was no longer necessary

SabrinaJewson commented:
This appears to cause hangs on some inputs. To reproduce:

1. Run `curl -O https://html.spec.whatwg.org/entities.json`
2. Put the following in `Cargo.toml`:

```toml
[package]
name = "repro"
version = "0.0.0"
edition = "2024"

[dependencies]
phf_codegen = "0.11.3"
serde = "1.0.218"
serde_json = "1.0.140"

[patch.crates-io]
phf_codegen   = { git = "https://github.com/conradludgate/rust-phf", rev = "7f55d882436b0c799fe63166309c2c096f51ecbe" }
phf_generator = { git = "https://github.com/conradludgate/rust-phf", rev = "7f55d882436b0c799fe63166309c2c096f51ecbe" }
phf_shared    = { git = "https://github.com/conradludgate/rust-phf", rev = "7f55d882436b0c799fe63166309c2c096f51ecbe" }
```
3. Put the following in `main.rs`:

```rs
fn main() {
    let json = fs::read("entities.json").unwrap();
    let entities = serde_json::from_slice::<HashMap<String, IgnoredAny>>(&json).unwrap();

    let mut map = phf_codegen::Map::new();
    for name in entities.keys() {
        if let Some(name) = name.strip_suffix(";") {
            map.entry(name.as_bytes(), "");
        }
    }
    map.build();
}

use serde::de::IgnoredAny;
use std::collections::HashMap;
use std::fs;
```

For me, `cargo run` will now hang (or maybe it’s just slow – either way, commenting out the `patch` section, or using a commit from before foldhash was implemented, instantly fixes the problem).

orlp commented:
There are two issues in foldhash at play here:

1. There is a small length extension attack possible that this alerted me to. I should incorporate the length into the hash mixing (despite Rust forcibly making me hash a `0xff` byte to stop length extension attacks) because I overlap inputs for speed. Currently `&iiint` and `&iiiint` always collide regardless of seed. 

2. Foldhash's `FixedState::with_seed` doesn't let you specify the full seed. It lets you change output order, but not always collision characteristics. It is much more rare, but possible that you find two strings which collide regardless of seed when using `FixedState::with_seed`, as the collisions is parametrized by the `FixedGlobalSeed` which is always hardcoded to be the same. This also affects `RandomState`, but to a much lesser extent as it regenerates the global seed every time.

orlp commented:
I've fixed both issue on the master branch of foldhash.

conradludgate commented:
I'll _fold_ in those changes soon, thanks!

CrazyboyQCD commented:
`b"zwj"` returns value but `b"zwnj"` returns `None`.
```rust
static ENTITIES_MAP: phf::Map<&'static [u8], &'static str> = phf_map! {
    b"zwj" => "0",
    b"zwnj" => "1",
};
```

orlp commented:
@CrazyboyQCD I can't reproduce. Did you make sure to patch all the involved dependencies?

```toml
[dependencies]
phf = { version = "0.11.3", features = ["macros"] }

[patch.crates-io]
phf_codegen   = { git = "https://github.com/conradludgate/rust-phf", rev = "b6bba72" }
phf_generator = { git = "https://github.com/conradludgate/rust-phf", rev = "b6bba72" }
phf_shared    = { git = "https://github.com/conradludgate/rust-phf", rev = "b6bba72" }
phf           = { git = "https://github.com/conradludgate/rust-phf", rev = "b6bba72" }
```

CrazyboyQCD commented:
@orlp 
My config is here, but with your config it works as expected:
```toml
[dependencies]
phf = { git = "https://github.com/conradludgate/rust-phf" , rev = "b6bba7216937ad4808c1a7097017aa88551a4463", features = ["macros"]}
```

orlp commented:
@CrazyboyQCD That only overrides `phf` itself, not the dependencies it has which generate the phf table. Phf will always catastrophically fail if there is a version mismatch between the generation algorithm and lookup algorithm.

conradludgate commented:
Looks like this will break `tokio-postgres`, cc @sfackler.

https://github.com/sfackler/rust-postgres/blob/490c4852492bb0adf63a83e3c4e8b2afc9e6bf54/codegen/src/sqlstate.rs#L139-L152

Since this codegen uses the phf version in the repo lockfile, which isn't guaranteed to match the phf version eventually used for the lookup.

Since you are a maintainer of both crates, I guess it would be interesting if you have opinions on whether:
1. A stability guarantee should be documented (which is already broken since unicase updated the `Hash` impl)
2. `tokio-postgres` should be using a build.rs or macro forms.

JohnTitor commented on phf_shared/src/foldhash.rs:
```diff
@@ -0,0 +1,258 @@
+// Copyright (c) 2024 Orson Peters
+// Distributed with permission under the zlib license.
+// Minor modifications made to remove any platform dependant code.
```
I'm not sure but does this affect our current license, MIT? I'd like to remain MIT as changing license is a hassle in general.

nicoburns replied:
Why does PR inline the foldhash implementation rather than depending on the `foldhash` crate? That would solve the licensing issue...
